database/sinkdb/sinkdb.go

```diff
@@ -0,0 +1,132 @@
+// Package sinkdb provides a strongly consistent key-value store.
+package sinkdb
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/golang/protobuf/proto"
+
+	"chain/database/raft"
+	"chain/database/sinkdb/internal/sinkpb"
+)
+
+// Open initializes the key-value store and returns a database handle.
+func Open(laddr, dir, bootURL string, httpClient *http.Client, useTLS bool) (*DB, error) {
+	state := newState()
+	sv, err := raft.Start(laddr, dir, bootURL, httpClient, useTLS, state)
+	if err != nil {
+		return nil, err
+	}
+	db := &DB{state: state, raft: sv}
+	return db, nil
+}
+
+// DB provides access to an opened kv store.
+type DB struct {
+	state *state
+	raft  *raft.Service
+}
+
+type Op struct {
+	err     error
+	conds   []*sinkpb.Cond
+	effects []*sinkpb.Op
+}
+
+// IfNotExists encodes a conditional to make an instruction
+// successful only if the provided key does not exist.
+func IfNotExists(key string) (op Op) {
+	op.conds = append(op.conds, &sinkpb.Cond{
+		Type: sinkpb.Cond_NOT_KEY_EXISTS,
+		Key:  key,
+	})
+	return op
+}
+
+// Delete encodes a delete operation for key.
+func Delete(key string) (op Op) {
+	op.effects = append(op.effects, &sinkpb.Op{
+		Type: sinkpb.Op_DELETE,
+		Key:  key,
+	})
+	return op
+}
+
+// Set encodes a set operation setting key to value.
+func Set(key string, value proto.Message) (op Op) {
+	encodedValue, err := proto.Marshal(value)
+	if err != nil {
+		op.err = err
+		return op
+	}
+
+	op.effects = append(op.effects, &sinkpb.Op{
+		Type:  sinkpb.Op_SET,
+		Key:   key,
+		Value: encodedValue,
+	})
+	return op
+}
+
+// Exec executes the provided operations. If all of the provided conditionals
+// are met, all of the provided effects are applied atomically.
+func (db *DB) Exec(ctx context.Context, ops ...Op) error {
```

Comment:
Does this function get us enough atomicity to do things like migrate multiple tables from pg? (e.g. right now we migrate config and access tokens separately, but that should be atomic)